jwst/cube_build/src/cube_dq_utils.c

```diff
@@ -0,0 +1,647 @@
+// This is a library of routine to set the DQ plane of the IFU cube
+
+#include <stdlib.h>
+#include <math.h>
+#include <stdio.h>
+#include <Python.h>
+#include <stdbool.h>
+
+// routines used from cube_utils.c
+
+extern double sh_find_overlap(const double xcenter, const double ycenter, 
+			      const double xlength, const double ylength,
+			      double xPixelCorner[],double yPixelCorner[]);
+
+
+//________________________________________________________________________________
+// allocate the memory for the spaxel DQ array
+
+int mem_alloc_dq(int nelem, int **idqv) {
+    
+    const char *msg = "Couldn't allocate memory for output arrays.";
+
+    if (!(*idqv = (int*)calloc(nelem, sizeof(int)))) {
+      PyErr_SetString(PyExc_MemoryError, msg);
+      return 1;
+    }
+
+    return 0;
+}
+
+
+//________________________________________________________________________________
+// Routine for MIRI DQ plane assignment
+
+int corner_wave_plane_miri(int w, int start_region, int end_region,
+			   double roiw_ave,
+			   double *zc,
+			   double *coord1, double *coord2, double *wave,
+			   double *sliceno,
+			   long ncube, int npt,
+			   double *corner1, double *corner2, double *corner3, double *corner4) {
+  /* 
+     For wavelength plane determine the corners (in xi,eta) of the FOV for MIRI
+     Use the 2 extreme slices set by start_region and end_region to define the FOV of the wavelength plane FOV
+     Using the min and max coordinates for the extent of the slice on the sky for these two slice - set the
+     corners of the FOV. 
+
+     w : wavelength plane
+     start_region : starting slice # for channel (slice # in nirspec) 
+     end_region : ending slice # for chanel      (slice # in nirspec)
+     roiw_ave : average roiw for all wavelengths
+     zc : array of wavelenths
+     coord1 : point cloud xi values
+     coord2 : point cloud eta values
+     wave : point cloud wavelength values
+     sliceno : point cloud slice no
+     ncube : number of cube values
+     npt: number of point cloud elements
+     corner1 : xi, eta of corner 1
+     corner2 : xi, eta of corner 2
+     corner3 : xi, eta of corner 3
+     corner4 : xi, eta of corner 4
+   */
+
+  int ipt, slice, c1_use;
+  double wave_distance;
+  float c11, c21, c12, c22, length_c1_start, length_c2_start;
+  int status = 0; 
+  int ic1_start_min = -1;
+  int ic1_start_max = -1;
+  int ic2_start_min = -1;
+  int ic2_start_max = -1;
+
+  float c1_start_min = 10000;
+  float c2_start_min = 10000;
+  float c1_start_max = -10000;
+  float c2_start_max = -10000;
+
+  int ic1_end_min = -1;
+  int ic1_end_max = -1;
+  int ic2_end_min = -1;
+  int ic2_end_max = -1;
+
+  float c1_end_min = 10000;
+  float c2_end_min = 10000;
+  float c1_end_max = -10000;
+  float c2_end_max = -10000;
```

Comment:
For now I am going to leave it. It is not as simple a change as the other ones because the values get set based wave_distance  and slice. But I will keep it as an area for improvement